### PR TITLE
make login, project, and discovery work against kube with RBAC enabled

### PR DESCRIPTION
**Pull request by deads2k**
When RBAC is enabled (see https://github.com/kubernetes/kubernetes/pull/34619), we have to tolerate 403s in addition to 404s.

@pweil- you'll probably need this if you want the tooling to work nicely (and it makes a difference)

### CODE REVIEW

**Comment by deads2k** [comment]
@fabianofranz @juanvallejo ping.

**Comment by juanvallejo** [comment]
@fabianofranz LGTM

**Comment by fabianofranz** [comment]
LGTM, @deads2k merge at will.

**Comment by deads2k** [comment]
[merge]

**Comment by deads2k** [comment]
yum re[test] re[merge]

**Comment by deads2k** [comment]
re[test]

**Comment by deads2k** [comment]
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/10125/ failed on https://github.com/openshift/origin/issues/11315
  re[test]

**Comment by deads2k** [comment]
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/10197/ failed on https://github.com/openshift/origin/issues/11315

re[test] re[merge]

**Comment by deads2k** [comment]
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/10265/ on https://github.com/openshift/origin/issues/11315

re[test] re[merge]

**Comment by deads2k** [comment]
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/10321/ on https://github.com/openshift/origin/issues/11315

re[test] re[merge

**Comment by deads2k** [comment]
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/10413/ on #11315

re[test] re[merge]

**Comment by deads2k** [comment]
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/10553/ on #11094

re[test] re[merge]

**Comment by deads2k** [comment]
@stevekuznetsov does networking ever work?  I keep getting #11315

**Comment by stevekuznetsov** [comment]
Unclear -- I don't have a good answer for you.

**Comment by deads2k** [comment]
@mfojtik @smarterclayton do we we want to force this one into 1.4?

**Comment by smarterclayton** [comment]
[merge]

**Comment by deads2k** [comment]
@openshift/networking I'm going to guess that somehow this broke the networking test.  Can you help me figure out how?

**Comment by danwinship** [comment]
tagging @marun since this is dind-related

If I check out that branch and do `make clean; ./test/extended/networking.sh`, I see the same deployment failure.

Doing `. dind-nettest.sh; oc get nodes` from another shell while it is waiting shows that the nodes are created and ready. But stracing the dind script shows that when it runs "oc get nodes ...", it gets back:

    the server doesn't have a resource type "nodes"

???

File: pkg/cmd/util/clientcmd/negotiate.go [hunk 1]
```diff
@@ -36,7 +36,7 @@ func negotiateVersion(client *kclient.Client, config *restclient.Config, request
 	// Get server versions
 	serverGVs, err := serverAPIVersions(client, "/oapi")
 	if err != nil {
-		if errors.IsNotFound(err) {
+		if errors.IsNotFound(err) || errors.IsForbidden(err) {
 			glog.V(4).Infof("Server path /oapi was not found, returning the requested group version %v", preferredGV)
 			return preferredGV, nil
```
Comment (danwinship) [inline pkg/cmd/util/clientcmd/negotiate.go]: This is the change that breaks the tests... Is there some other reason we could be getting a 403 here in some circumstances?

**Comment by smarterclayton** [comment]
Why is a race with initial cache priming crazy?

File: pkg/cmd/util/clientcmd/negotiate.go (hunk 1 above)
Comment (danwinship) [inline pkg/cmd/util/clientcmd/negotiate.go]: And this is against an openshift server

**Comment by marun** [comment]
Has this patch been tested against a non-dind multinode cluster?

**Comment by marun** [comment]
I'm seeing the failure consistently after rebasing the PR, which is how it is being tested in CI.

**Comment by marun** [comment]
Assuming stale cache in ``~/.kube`` is the cause of the CI failure, consider updating ``test/extended/networking.sh`` to clear it and seeing if that allows the job to pass.

**Comment by liggitt** [comment]
> consider updating test/extended/networking.sh to clear it and seeing if that allows the job to pass

this confuses me... the cache shouldn't be storing anything in error cases, right?

**Comment by deads2k** [comment]
> this confuses me... the cache shouldn't be storing anything in error cases, right?

riddle me this.  How does this pull produce this?

https://paste.fedoraproject.org/466900/14779326/

**Comment by deads2k** [comment]
It almost looks like networking is starting an openshift server that does not run the kubernetes API server, but instead tries to proxy it and somehow the proxy isn't proxying the discovery doc.  That's insane though.  Running an openshift API server like that won't work and I don't *think* this code will react differently in that case than the old code would have.

**Comment by danwinship** [comment]
> It almost looks like networking is starting an openshift server that does not run the kubernetes API server

dind runs openshift in several different modes while setting up the cluster. eg, first it runs `openshift admin ca create-master-certs ...`, then `openshift start master --write-config ...`, etc. (see images/dind/master/). And the "oc get nodes" loop to test if the cluster is ready starts before those config steps finish. So if one of those returns bad data, and oc caches it, then...

**Comment by deads2k** [comment]
> dind runs openshift in several different modes while setting up the cluster. eg, first it runs openshift admin ca create-master-certs ..., then openshift start master --write-config ..., etc. (see images/dind/master/). And the "oc get nodes" loop to test if the cluster is ready starts before those config steps finish. So if one of those returns bad data, and oc caches it, then...

Thing is, none of these are server-side changes, so the discovery information being served is identical and whatever command eventually saved the empty discovery doc, got that back from the server it queried.  But, every other component that starts a master is serving "normal" discovery information from the discovery endpoints.

**Comment by danwinship** [comment]
I meant, they run before the master does, maybe they're binding to port 8443 and serving bogus data. But it looks like they don't.

Is there any chance the master itself could be returning bad answers briefly at startup? Like, does the startup code do something like:

    kube.StartMasterHTTPServer()
    openshift.OhBTWHandleOpenShiftURLsToo()

**Comment by deads2k** [comment]
> I meant, they run before the master does, maybe they're binding to port 8443 and serving bogus data. But it looks like they don't.
> 
> Is there any chance the master itself could be returning bad answers briefly at startup? Like, does the startup code do something like:

Even so, this doesn't change that behavior one way or the other.  Are you guys running an `oc login --token` or `oc project` command somewhere on a loop?  That might start succeeding sooner than usual.

**Comment by stevekuznetsov** [comment]
@danwinship there is one handler chain for the server that serves everything, described in [`MasterConfig.Run()`](https://github.com/openshift/origin/blob/master/pkg/cmd/server/origin/master.go#L168-L360)

**Comment by deads2k** [comment]
> Is there any chance the master itself could be returning bad answers briefly at startup? Like, does the startup code do something like:

@danwinship That's a good theory.  Can you link me Where is the `oc get nodes` is done so I can switch it to a health check first?

**Comment by danwinship** [comment]
`wait-for-cluster` in `hack/dind-cluster.sh`

**Comment by marun** [comment]
I don't see how the master could be returning bad answers briefly at startup, because if I am deploying a dind cluster after having removed ``~/kube``, everything works fine.  ``oc get nodes`` works just fine, too.  It's only if I wait (like 5m) and then try running ``oc get nodes`` that I'm seeing the failure indicated by fpaste.

**Comment by marun** [comment]
I don't think this issue is specific to dind deployment, and that merging should wait until the networking job is passing.

**Comment by deads2k** [comment]
> I don't think this issue is specific to dind deployment, and that merging should wait until the networking job is passing.


I haven't proposed forcing it.

I'm not familiar with the job though.  It would be nice to have a smaller, faster reproducer to help track this down.  Is there a flag to prevent tear down and then scripts to debug using various kubeconfigs?  The differences in this job make it harder to jump in and debug.

**Comment by marun** [comment]
@deads2k ``hack/dind-cluster.sh start -r`` reproduces the issue for me.  One the dind images have been built, deploying a cluster takes 20-30s.  The only requirement is linux running recent (> 1.10) docker.

My comment about merge was in response to the bot trying to merge after your push.  I realize now that the consistent job failure means it can't succeed.

**Comment by deads2k** [comment]
> @deads2k hack/dind-cluster.sh start -r reproduces the issue for me. Once the dind images have been built, deploying a cluster takes 20-30s. The only requirement is linux running recent (> 1.10) docker.

It doesn't reproduce it for me.  That command always seems to work.

**Comment by deads2k** [comment]
> @deads2k hack/dind-cluster.sh start -r reproduces the issue for me. Once the dind images have been built, deploying a cluster takes 20-30s. The only requirement is linux running recent (> 1.10) docker.
> It doesn't reproduce it for me. That command always seems to work.

@marun can you reproduce this reliably on an AWS instance I could use to try to diagnose it.  The pastebin you made doesn't seem to appear in the jenkins test run (near as I can tell) and every other master we test it against works.  This really looks like its specific to networking test provisioning somehow, I can't seem to make it fail locally, and I can't figure out where the failure happens and is logged for instrumenting while running jenkins.

File: hack/dind-cluster.sh [hunk 2]
```diff
@@ -221,7 +221,7 @@ function wait-for-cluster() {
   oc="$(os::build::find-binary oc)"
 
   # wait for healthz to report ok before trying to get nodes
-  os::util::wait-for-condition "ok" "${oc} get --config=\"${kubeconfig}\" --raw=/healthz" "120"
```
Comment (stevekuznetsov) [inline hack/dind-cluster.sh]: The function in question claims it uses `eval` but in fact does this:

```sh
  while ! $(${condition}); do
```

This means that yes, these quotes were incorrect and the actual value of `--config` that is passed to `oc get` is `"${kubeconfig}"`, literal quotes and all. @marun this type of thing is why I feel so strongly about _not_ having the `provision_util.sh` file, since we spent a lot of time and effort getting it right the first time in `os::cmd`.

File: hack/dind-cluster.sh [hunk 3]
```diff
@@ -249,7 +249,7 @@ EOF
   # Remove formatting before use
   template="$(echo "${template}" | tr -d '\n' | sed -e 's/} \+/}/g')"
   local count
-  count="$("${oc}" --config="${kubeconfig}" get nodes \
+  count="$("${oc}" --config=${kubeconfig} get nodes \
```
Comment (stevekuznetsov) [inline hack/dind-cluster.sh]: I can't get this one to fail for me, and these quotes look fine. Revert this change, please.

File: hack/dind-cluster.sh (hunk 2 above)
Comment (marun) [inline hack/dind-cluster.sh]: ``provision_util.sh`` isn't used here.  That's legacy and only used by vagrant.  I'm assuming you meant ``images/dind/node/openshift-dind-lib.sh``, which needs to be a separate file so it can be distributed in the dind image.  I'm happy to have ``os::cmd`` take over responsibility for this use case, but it would have to be copied into the image file for distribution regardless.

Comment (stevekuznetsov) [inline hack/dind-cluster.sh]: Ah, whatever checkout of Origin I'm in right now has the `os::util::wait-for-condition` function in `provision-util.sh`.

**Comment by deads2k** [comment]
and it worked.  I'll squash down and eliminate the "extra" unquoting.

File: hack/dind-cluster.sh [hunk 4]
```diff
@@ -220,6 +220,9 @@ function wait-for-cluster() {
   local oc
   oc="$(os::build::find-binary oc)"
 
+  # wait for healthz to report ok before trying to get nodes
```
Comment (deads2k) [inline hack/dind-cluster.sh]: > I don't see why this change would be necessary. Consider removing.

I think its reasonable to wait until the API server is ready before making real requests to it. This mirrors what other e2e tests do.

**Comment by ncdc** [comment]
@deads2k congrats, tests passed
